### PR TITLE
PYIC-2706: use featureset from header to get config

### PR DESCRIPTION
**Pull request by shivanshuit914**
## Proposed changes

- Passing featureset to config service from all lambda using setter method.

### What changed

- Added extra call to config service to set featureset.

### Issue tracking

- [PYIC-2706](https://govukverify.atlassian.net/browse/PYIC-2706)

### Environment variables or secrets

- [x] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

[PYIC-2706]: https://govukverify.atlassian.net/browse/PYIC-2706?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by lwebb-iig — DISMISSED** [review]

**Review by lwebb-iig — DISMISSED** [review]
LGTM - might want to get one of the refactor guys to look at this as well

File: lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
 import uk.gov.di.ipv.core.library.exceptions.HttpResponseExceptionWithErrorBody;
 import uk.gov.di.ipv.core.library.exceptions.SqsException;
 import uk.gov.di.ipv.core.library.helpers.LogHelper;
+import uk.gov.di.ipv.core.library.helpers.RequestHelper;
```
Comment (therealslimsladey) [inline lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:35]: In `BuildClientOauthResponse`, `getFeatureSet` is an explicit static import. Wouldn't it make sense to do that here too for consistency?

File: lib/src/main/java/uk/gov/di/ipv/core/library/helpers/LogHelper.java [hunk 2]
```diff
@@ -84,6 +84,10 @@ public static void attachIpvSessionIdToLogs(String sessionId) {
         attachFieldToLogs(LogField.LOG_IPV_SESSION_ID, sessionId);
     }
 
+    public static void attachFeatureSetToLogs(String sessionId) {
```
Comment (rpayal) [inline lib/src/main/java/uk/gov/di/ipv/core/library/helpers/LogHelper.java]: good to have param name as featureSet.

File: lib/src/main/java/uk/gov/di/ipv/core/library/helpers/RequestHelper.java [hunk 3]
```diff
@@ -137,15 +137,16 @@ public static String getIpvSessionId(JourneyRequest request, boolean allowNull)
...
-            LOGGER.warn("{} not present in header", FEATURE_SET_HEADER);
-            return "default";
-        }
+    public static String getFeatureSet(JourneyRequest request) {
+        String featureSet = request.getFeatureSet();
+        LogHelper.attachFeatureSetToLogs(featureSet);
         return featureSet;
     }
 
+    public static String getFeatureSet(Map<String, String> headers) {
```
Comment (rpayal) [inline lib/src/main/java/uk/gov/di/ipv/core/library/helpers/RequestHelper.java]: this method can be private.

File: lib/src/main/java/uk/gov/di/ipv/core/library/statemachine/BaseJourneyLambda.java [hunk 4]
```diff
@@ -48,9 +48,12 @@ public Map<String, Object> handleRequest(Map<String, Object> event, Context cont
                     OBJECT_MAPPER.convertValue(event, APIGatewayProxyRequestEvent.class);
 
             var clientOAuthSessionId = RequestHelper.getClientOAuthSessionId(request);
+            var ipvSessionId = getIpvSessionId(request);
+            var ipAddress = getIpAddress(request);
+            var featureSet = RequestHelper.getFeatureSet(request);
```
Comment (rpayal) [inline lib/src/main/java/uk/gov/di/ipv/core/library/statemachine/BaseJourneyLambda.java]: can have static import here.

File: lambdas/retrieve-cri-oauth-access-token/src/main/java/uk/gov/di/ipv/core/retrievecrioauthaccesstoken/RetrieveCriOauthAccessTokenHandler.java [hunk 5]
```diff
@@ -87,6 +87,8 @@ public Map<String, Object> handleRequest(Map<String, String> input, Context cont
         String credentialIssuerId = null;
 
         String ipAddress = StepFunctionHelpers.getIpAddress(input);
+        String featureSet = StepFunctionHelpers.getFeatureSet(input);
```
Comment (therealslimsladey) [inline lambdas/retrieve-cri-oauth-access-token/src/main/java/uk/gov/di/ipv/core/retrievecrioauthaccesstoken/RetrieveCriOauthAccessTokenHandler.java:90]: RequestHelpers.getFeatureSet attaches the featureSet to the logs. StepFunctionHelpers.getFeatureSet doesn't - it probably should?

File: lib/src/main/java/uk/gov/di/ipv/core/library/helpers/LogHelper.java (hunk 2 above)
Comment (shivanshuit914) [reply lib/src/main/java/uk/gov/di/ipv/core/library/helpers/LogHelper.java]: Renamed it, thanks

File: lib/src/main/java/uk/gov/di/ipv/core/library/statemachine/BaseJourneyLambda.java (hunk 4 above)
Comment (shivanshuit914) [reply lib/src/main/java/uk/gov/di/ipv/core/library/statemachine/BaseJourneyLambda.java]: done.

File: lib/src/main/java/uk/gov/di/ipv/core/library/helpers/RequestHelper.java (hunk 3 above)
Comment (shivanshuit914) [reply lib/src/main/java/uk/gov/di/ipv/core/library/helpers/RequestHelper.java]: done

File: lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java [hunk 6]
```diff
@@ -114,6 +114,8 @@ public Map<String, Object> handleRequest(CriCallbackRequest callbackRequest, Con
         try {
             String ipvSessionId = callbackRequest.getIpvSessionId();
             String criOAuthSessionId = callbackRequest.getState();
+            String featureSet = callbackRequest.getFeatureSet();
```
Comment (therealslimsladey) [inline lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java:117]: We won't get the featureSet into the logging context through the Lombok getter

File: lib/src/main/java/uk/gov/di/ipv/core/library/helpers/RequestHelper.java [hunk 7]
```diff
@@ -137,15 +137,16 @@ public static String getIpvSessionId(JourneyRequest request, boolean allowNull)
...
+        return RequestHelper.getHeaderByKey(headers, FEATURE_SET_HEADER);
```
Comment (therealslimsladey) [inline lib/src/main/java/uk/gov/di/ipv/core/library/helpers/RequestHelper.java:147]: Should be attaching the feature set to the log context - or better, the ConfigService setter may be the best place for this. Let's discuss?

**Review by lwebb-iig — APPROVED** [review]

**Comment by therealslimsladey** [comment]
BaseJourneyLambda test coverage is low

**Comment by shivanshuit914** [comment]
> BaseJourneyLambda test coverage is low

Yes I am trying to add unit tests, there is no test at the moment may be because of abstract class.

**Pull request merged**